Report live value of decentralization parameter d.

jonathanknowles opened the pull request:
# Issue Number

#1693 

# Overview

This PR:

- [x] Introduces the `DecentralizationLevel` type (based on the `Percentage` type).

    _Greater_ values indicate _greater_ levels of decentralization, as follows:

    | Value | Meaning |
    | --: | :-- |
    | 0% | Network is completely _federalized_ |
    | 100% | Network is completely _decentralized_ |

- [x] Adjusts the **wallet API server** to convert from values of type `DecentralizationLevel` to values of type `Quantity "percent" Percentage`.

- [x] Adjusts the **wallet layer** to store a value of type `ProtocolParameters` instead of `TxParameters`. (Note that `ProtocolParameters` _includes_ `TxParameters`.)

- [x] Adjusts the **network layer** to provide a `getProtocolParameters` function instead of `getTxParameters`.

- [x] Adjusts the **database layer** to use a `protocol_parameters` table instead of a `tx_parameters` table. (Note that `protocol_parameters` is a superset of `tx_parameters`.)

- [x] Adjusts the **chain-following code** to extract the live value of `d`, in the same way as we already do for transaction-related protocol parameters.

- [x] Adds a **conversion function** to convert from `d` parameter values to values of type `DecentralizationLevel`.

    This function performs the following mapping:
    ```
    {1 → 0%, 0.9 → 10%, 0.8 → 20%, ..., 0 → 100%}
    ```

# Comments

Starting a wallet with this PR applied will result in the creation of a new `protocol_parameters` table, if one does not already exist. (We rely on the automatic migration mechanism to create the table, and the network layer to add rows to the table.)

The redundant `tx_parameters` table, if it exists, will be ignored, but _not_ deleted.

In a future PR, we can add a manual migration step that removes the `tx_parameters` table.

See Issue https://github.com/input-output-hk/cardano-wallet/issues/1727.

jonathanknowles committed:
Adjust formatting to meet coding standards.

github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles committed:
Generalize `TxParameters` to `ProtocolParameters`.

Generalize `TxParameters` to `ProtocolParameters` in the wallet, network
and DB layers.

jonathanknowles committed:
Introduce type `DecentralizationLevel`.

jonathanknowles committed:
Extract the live decentralization parameter value.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -345,18 +346,36 @@ fromMaxTxSize :: Natural -> Quantity "byte" Word16
+            (Quantity (fromIntegral (SL._minfeeA pp)))
+            (Quantity 0) -- TODO: it's not as simple as this?
+        , getTxMaxSize = fromMaxTxSize $ SL._maxTxSize pp
+        }
```
Perhaps we should also extract a `txParametersFromPParams` function here.

jonathanknowles replied:
Fixed in 53352b61478ce658986a596b03d8691c2b80120f.

jonathanknowles committed:
Extract out function `txParametersFromPParams`.

rvl commented on lib/byron/src/Cardano/Wallet/Byron/Network.hs:
```diff
@@ -597,8 +603,8 @@ instance ToText NetworkLayerLog where
-        MsgTxParameters params -> T.unwords
-            [ "TxParams for tip are:"
+        MsgProtocolParameters params -> T.unwords
+            [ "ProtocolParams for tip are:"
```
How does this log message look? Would it be better as a multiline text (`T.unwords` → `T.unlines`)?

jonathanknowles replied:
> How does this log message look? Would it be better as a multiline text (`T.unwords` → `T.unlines`)?

Currently it looks like this:

```
ProtocolParams for tip are: [Decentralization level: 0.00%, Transaction parameters: [Fee policy: 14.0 + 42.0x + 5.0y, Tx max size: 8192]]
```

Using `T.unlines` makes it look like this:
```
ProtocolParams for tip are:
[Decentralization level: 0.00%, Transaction parameters: [Fee policy: 14.0 + 42.0x + 5.0y, Tx max size: 8192]]
```

I'll try to make some further improvements.

jonathanknowles replied:
With commit 4de805b147e5626614c36d12fa65958db7bb9220, the log output now looks like this:

```
Protocol parameters for tip are:
 Decentralization level: 0.00%
 Transaction parameters: [Fee policy: 14.0 + 42.0x + 5.0y, Tx max size: 8192]
```

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -345,8 +346,50 @@ fromMaxTxSize :: Natural -> Quantity "byte" Word16
+        $ SL.intervalValue d
+  where
+    d = SL._d pp
+    reportInvalidValue = error $ mconcat
```
What about adding a `HasCallStack` constraint to this function and its callers?

jonathanknowles replied:
Fixed in 4bc118bf324f63f183be82c85731e4042d3c87f5.

rvl reviewed (approved):
Looks good so far

jonathanknowles committed:
Add `HasCallStack` constraint.

Add `HasCallStack` constraint to function `decentralizationLevelFromPParams`
and all functions that call it directly.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1725#discussion_r435283572

jonathanknowles committed:
Improve the formatting of `MsgProtocolParameters`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1725#discussion_r435278582

jonathanknowles committed:
Introduce function `invertUnitInterval`.

jonathanknowles committed:
Add comment to explain inversion of d parameter.

jonathanknowles committed:
Add unit test for `decentralizationLevelFromPParams`.

rvl reviewed (approved):
Great! :+1:
